state-chain/pallets/cf-swapping/src/lib.rs

```diff
@@ -1246,154 +1220,248 @@ pub mod pallet {
-									swap_request_id,
-									asset: swap.output_asset(),
-									amount: output_amount,
-									reason: err.into(),
-								});
-							},
-						};
-					},
-					SwapRequestState::NetworkFee => {
-						Self::deposit_event(Event::<T>::SwapRequestCompleted { swap_request_id });
-						if swap.output_asset() == Asset::Flip {
-							FlipToBurn::<T>::mutate(|total| {
-								total.saturating_accrue(output_amount);
-							});
-						} else {
-							log_or_panic!(
-								"NetworkFee burning should not be in asset: {:?}",
-								swap.output_asset()
-							);
-						}
-					},
-					SwapRequestState::IngressEgressFee => {
-						Self::deposit_event(Event::<T>::SwapRequestCompleted { swap_request_id });
+					};
 
-						if swap.output_asset() ==
-							ForeignChain::from(swap.output_asset()).gas_asset()
-						{
-							T::IngressEgressFeeHandler::accrue_withheld_fee(
-								swap.output_asset(),
-								output_amount,
+					if is_gas_swap {
+						*gas_swap_state = GasSwapState::OutputReady { gas_budget: output_amount };
+					} else {
+						// The executed swap must be for the principal amount:
+						dca_state.accumulated_output_amount += output_amount;
+
+						// Schedule any remaining DCA chunks:
+						dca_state.scheduled_chunk_swap_id = if dca_state.remaining_chunks > 0 {
+							let chunk_input_amount = dca_state.prepare_next_chunk();
+
+							let swap_id = Self::schedule_swap(
+								request.input_asset,
+								request.output_asset,
+								chunk_input_amount,
+								swap.swap.refund_params.clone(),
+								SwapType::CcmPrincipal,
+								swap_request_id,
 							);
+
+							Some(swap_id)
 						} else {
-							log_or_panic!(
-								"IngressEgressFee swap should not be to non-gas asset: {:?}",
-								swap.output_asset()
+							None
+						};
+
+						// See if we still need to schedule the gas swap
+						if let GasSwapState::ToBeScheduled { gas_budget, other_gas_asset } =
+							gas_swap_state
+						{
+							*gas_swap_state = Self::schedule_ccm_gas_swap(
+								swap_request_id,
+								request.input_asset,
+								*other_gas_asset,
+								*gas_budget,
 							);
 						}
-					},
-				}
+					}
+
+					// Process outcome if there aren't any more scheduled swaps,
+					// otherwise put the request state back:
+					if let (GasSwapState::OutputReady { gas_budget }, None) =
+						(gas_swap_state, dca_state.scheduled_chunk_swap_id)
+					{
+						Self::egress_for_swap(
+							swap_request_id,
+							dca_state.accumulated_output_amount,
+							request.output_asset,
+							output_address.clone(),
+							Some((ccm_deposit_metadata.clone(), *gas_budget)),
+							false, /* refund */
+						);
+
+						true
+					} else {
+						false
+					}
+				},
+				SwapRequestState::Regular { output_address, dca_state } => {
+					if dca_state.remaining_chunks > 0 {
+						let chunk_input_amount = dca_state.prepare_next_chunk();
+						dca_state.accumulated_output_amount += output_amount;
```

Comment:
We only add the output amount if the remaining chunks > 0. Is this intentional? In the ccm branch this line is outside the if clause. 

Seems like a good candidate for deduplication.

Reply:
This part of the code uses if/else, which I think should be preferred over two separate if statements. The CCM case is more complex (two swaps can be executed concurrently), so a separate if statement was needed.

Reply:
But for example here, I can see that we don't update/assign the scheduled_chunk_swap_id. Maybe it's not significant, given that we only use it during ccm(?), but I can see how it will cause confusion. 

Surely when there is a swap execution, we *always* want to update the dca_state in the same way? So why can't we move this outside of the checks and do it consistently for every type of swap?

Reply:
> given that we only use it during ccm(?)

The intention was to also use it for regular swaps, just as a sanity check at least. Will update.